services/leader/client.go

```diff
@@ -0,0 +1,220 @@
+package leader
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/leader/campaign"
+	"github.com/m3db/m3cluster/services/leader/election"
+
+	"golang.org/x/net/context"
+)
+
+// appended to elections with an empty string for electionID to make it easier
+// for user to debug etcd keys
+const defaultElectionID = "default"
+
+var (
+	// ErrNoLeader is returned when a call to Leader() is made to an election
+	// with no leader. We duplicate this error so the user doesn't have to
+	// import etcd's concurrency package in order to check the cause of the
+	// error.
+	ErrNoLeader = concurrency.ErrElectionNoLeader
+
+	// ErrCampaignInProgress is returned when a call to Campaign() is made while
+	// the caller is either already (a) campaigning or (b) the leader.
+	ErrCampaignInProgress = errors.New("campaign in progress")
+)
+
+// NB(mschalle): when an etcd leader failover occurs, all current leases have
+// their TTLs refreshed: https://github.com/coreos/etcd/issues/2660
+
+type client struct {
+	sync.RWMutex
+
+	ecl         *election.Client
+	opts        services.ElectionOptions
+	ctxCancel   context.CancelFunc
+	resignC     chan struct{}
+	campaigning uint32
+	closed      uint32
+}
+
+// newClient returns an instance of an client bound to a single election.
+func newClient(cli *clientv3.Client, opts Options, electionID string, ttl int) (*client, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	var sessionOpts []concurrency.SessionOption
+	if ttl != 0 {
+		sessionOpts = append(sessionOpts, concurrency.WithTTL(ttl))
+	}
+
+	pfx := electionPrefix(opts.ServiceID(), electionID)
+	ec, err := election.NewClient(cli, pfx, election.WithSessionOptions(sessionOpts...))
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		ecl:     ec,
+		opts:    opts.ElectionOpts(),
+		resignC: make(chan struct{}),
+	}, nil
+}
+
+func (c *client) campaign(opts services.CampaignOptions) (<-chan campaign.Status, error) {
+	if c.isClosed() {
+		return nil, errClientClosed
+	}
+
+	if !c.startCampaign() {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c.Lock()
```

Comment:
I thought we need it since multiple goroutines can simultaneously read in `resign()`?